opentelemetry/src/runtime.rs

```diff
@@ -0,0 +1,87 @@
+//! Provides an abstraction of several runtimes, which allows OpenTelemetry to work with any
+//! current or future runtime. There are currently builtin implementations for Tokio and async-std.
+
+use futures::{future::BoxFuture, stream::BoxStream};
+use std::time::Duration;
+
+/// A runtime in an abstraction of a (possibly async runtime) like tokio or async-std. It allows
+/// OpenTelemetry to work with any current and future runtime implementation.
+pub trait Runtime: Clone {
+    /// Create a [Stream][futures::Stream], which returns a new item every

```

Comment:
Did you consider using associated types for the return types here, or are you aiming to keep this object safe? If the latter, might be useful to make that explicit in a comment.

Reply:
Yeah, I was undecided about object safety. If that's not needed we can use associated types instead, which would remove some mapping in the functions.

Reply:
I'd maybe start with a smaller take that just adds the `Runtime` trait, its impls, and replaces the outer layer of methods that take the interminable number of parameters with it, and use associated types (equivalent to what's there now, just bundled up in a trait).

Reply:
(But on the whole, I really like this approach and think it's definitely better than what we have now -- even though I haven't reviewed it in full detail just yet.)

Reply:
Agree, this is a nice improvement! @frigus02 any issues you're still working through? or close to having this ready to review?

Reply:
Sorry the slow progress. The last week was quite busy for me. I plan on on working on this tomorrow and get it ready for review.

I'll try to split this into 2 things:

1. The runtime trait but without public API changes.
2. The changes from cargo features to Rust code.

This way things should be easier to review and we can also talk about the changes separately.

Reply:
Alright. I split this PR into two: #480 and #481